File: 0131-palindrome-partitioning/0131-palindrome-partitioning.cpp
class Solution {
public:
    vector<vector<string>>res;
    bool isPalindrome(string s,int start,int end)
    {
        while(start < end)
        {
            if(s[start] != s[end])
            {
                return false;
            }
            start++;
            end--;
        }
        return true;
    }
    void dfs(string s,int idx,vector<string>&arr)
    {
        if(idx == s.size())
        {
            res.push_back(arr);
            return;
        }
        
        for(int i=idx;i<s.size();i++)
        {
            if(isPalindrome(s, idx, i))
            {
                arr.push_back(s.substr(idx, i - idx + 1));
                dfs(s, i + 1, arr);
                arr.pop_back();
            }
        }
    }
    vector<vector<string>> partition(string s) {
        vector<string>arr;
        dfs(s,0,arr);
        return res;
    }
};

File: 0131-palindrome-partitioning/0131-palindrome-partitioning.py
class Solution:
    def solve(self,idx,path,s):
        if idx == len(s):
            self.res.append(path[:])
            return
        for end in range(idx,len(s)):
            if s[idx:end+1] == s[idx:end+1][::-1]:
                path.append(s[idx:end+1])
                self.solve(end+1,path,s)
                path.pop()
        
    def partition(self, s: str) -> List[List[str]]:
        self.res = []
        self.solve(0,[],s)
        return self.res